[distributed] add base processgroup::options

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53663 [distributed] add processgroup options as argument
* #54090 [distributed] add options field in ProcessGroupGloo/NCCL
* **#53662 [distributed] add base processgroup::options**

Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

Differential Revision: [D26968856](https://our.internmc.facebook.com/intern/diff/D26968856)

wanchaol committed:
[distributed] add base processgroup::options

Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add base processgroup::options"


Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1049,6 +1049,12 @@ that adds a prefix to each key inserted to the store.
+  auto processGroupOptions =
+      intrusive_ptr_class_<::c10d::ProcessGroup::Options>(
+          module, "ProcessGroupOptions")
+          .def_readwrite("backend", &::c10d::ProcessGroup::Options::backend);
```
It looks like the `10d::ProcessGroup::Options` also has a `timeout`, should we add that binding here as well?

wanchaol replied:
I thought I added it already lol, weird it doesn't trigger any failures

rohan-varma commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -164,6 +166,15 @@ class ProcessGroup : public torch::CustomClassHolder {
     std::function<void()> recordFunctionEndCallback_;
   };
 
+  struct Options : torch::CustomClassHolder {
```
Nit: would be good to add a comment about this class

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.hpp:
```diff
@@ -132,11 +132,15 @@ class ProcessGroupGloo : public ProcessGroup {
     explicit Options();
 
+    // return intrusive_ptr of the object
+    static c10::intrusive_ptr<Options> create() {
```
Is there a reason we've converted to always returning intrusive_ptr of the options instead of the regular struct as before?

wanchaol replied:
Mostly for TorchScript compliance, custom classes are better supported in TorchScript if we pass them by intrusive_ptr.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -982,9 +982,12 @@ void ProcessGroupNCCL::workEnqueue(
+
+ProcessGroupNCCL::Options::Options(
+    std::chrono::milliseconds timeout,
+    bool is_high_stream)
```
nit: `bool is_high_priority_stream`?

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -182,18 +182,21 @@ class ProcessGroupNCCL : public ProcessGroup {
-      return c10::make_intrusive<Options>();
+        std::chrono::milliseconds timeout =
+            std::chrono::milliseconds(kProcessGroupNCCLOpTimeoutMillis),
+        bool is_high_stream = false) {
```
ditto

rohan-varma reviewed (approved):
LGTM!

wanchaol committed:
Update on "[distributed] add base processgroup::options"


Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

Differential Revision: [D26968856](https://our.internmc.facebook.com/intern/diff/D26968856)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add base processgroup::options"


Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

Differential Revision: [D26968856](https://our.internmc.facebook.com/intern/diff/D26968856)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add base processgroup::options"


Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

Differential Revision: [D26968856](https://our.internmc.facebook.com/intern/diff/D26968856)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add base processgroup::options"


Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

Differential Revision: [D26968856](https://our.internmc.facebook.com/intern/diff/D26968856)

[ghstack-poisoned]

wanchaol committed:
Update on "[distributed] add base processgroup::options"


Add a base processgroup::options so that we can do inheritance and
provide
a universal option API in python

Differential Revision: [D26968856](https://our.internmc.facebook.com/intern/diff/D26968856)

[ghstack-poisoned]
